File: aoc24c/day10.c
// Advent of Code 2024
// Dr Lee A. Christie
//
// GitHub:   @leechristie
// Mastodon: @0x1ac@techhub.social
// Website:  leechristie.com

#include <ctype.h>
#include <stdio.h>

#include "days.h"
#include "grid.h"
#include "list.h"
#include "timing.h"

static int evaluate(const CharGrid * const grid, const int index, IntList * const visited_end_points, const char expected_symbol) {

    // invalid position
    if (index < 0)
        return 0;

    // wrong symbol
    const char actual_symbol = CharGrid_Get_ByIndex_OrDefault(grid, index, '\0');
    if (actual_symbol != expected_symbol)
        return 0;

    // found goal
    if (expected_symbol == '9') {
        IntList_SmallSet_Add(visited_end_points, index);
        return 1;
    }

    // not finished - check neighbours
    int rv = 0;
    rv += evaluate(grid, (int) CharGrid_Index_With_YXOffset(grid, index, -1, 0), visited_end_points, expected_symbol + 1);
    rv += evaluate(grid, (int) CharGrid_Index_With_YXOffset(grid, index, 0, 1), visited_end_points, expected_symbol + 1);
    rv += evaluate(grid, (int) CharGrid_Index_With_YXOffset(grid, index, 1, 0), visited_end_points, expected_symbol + 1);
    rv += evaluate(grid, (int) CharGrid_Index_With_YXOffset(grid, index, 0, -1), visited_end_points, expected_symbol + 1);
    return rv;

}

int day10() {

    start_timer();

    int part1 = 0;
    int part2 = 0;

    // read the grid
    FILE * file = fopen("input10.txt", "r");
    if (file == NULL) {
        printf("unable to open input file");
        return 1;
    }
    CharGrid grid;
    CharGrid_FromFile(&grid, file);
    fclose(file);

    // solve
    IntList visited_end_points;
    IntList_Init(&visited_end_points);
    for (int y = 0; y < (int) grid.height; y++) {
        for (int x = 0; x < (int) grid.width; x++) {
            int index = (int) CharGrid_YX_To_Index(&grid, y, x);
            part2 += evaluate(&grid, index, &visited_end_points, '0');
            part1 += (int) visited_end_points.length;
            IntList_Clear(&visited_end_points);
        }
    }

    // cleanup
    CharGrid_DeInit(&grid);
    IntList_DeInit(&visited_end_points);

    const double time = stop_timer();
    printf("Advent of Code 2024\n");
    printf("Day 10 - Hoof It\n");
    printf("Part 1: %d\n", part1);
    printf("Part 2: %d\n", part2);
    printf("Time Taken: %.6f s\n", time);
    return 0;

}

File: aoc24py/day10.py
# Advent of Code 2024
# Dr Lee A. Christie
#
# GitHub:   @leechristie
# Mastodon: @0x1ac@techhub.social
# Website:  leechristie.com


import time
from typing import Optional
from grid import CharacterGrid


def evaluate(grid: CharacterGrid, location: tuple[int, int], ordinal: int) -> tuple[int, set[tuple[int, int]]]:

    # wrong number
    here: Optional[str] = grid[location]
    if here is None or ord(here) != ordinal:
        return 0, set()

    # found goal
    if ordinal == ord('9'):
        return 1, {location}

    # check neighbours
    y, x = location
    trail_count = 0
    trail_ends = set()
    for n in [(y - 1, x), (y, x + 1), (y + 1, x), (y, x - 1)]:
        score, ends = evaluate(grid, n, ordinal + 1)
        trail_count += score
        trail_ends.update(ends)

    # return all number of trails and reachable trail ends
    return trail_count, trail_ends


def day10() -> None:

    start: float = time.perf_counter()

    part1 = 0
    part2 = 0

    grid = CharacterGrid.read_character_grid("input10.txt")

    for y in range(grid.height):
        for x in range(grid.width):
            location = y, x
            value = grid[location]
            if value == '0':
                count, ends = evaluate(grid, location, ord(value))
                part1 += len(ends)  # number of reachable ends
                part2 += count      # count of unique trails

    stop: float = time.perf_counter()

    print("Advent of Code 2024")
    print("Day 10 - Hoof It")
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")
    print(f"Time Taken: {stop-start:.6f} s")
